### PR TITLE
Add IDockerClientAuthenticationConfiguration; Clean up project.

### PR DESCRIPTION
**Pull request by HofmeisterAn**
- Removed the `Endpoint` property.
- Added `IDockerClientAuthenticationConfiguration`.
- Cleaned up some parts (code style, etc).
- Added `DockerClientEnvironmentAuthenticationConfiguration`.
- Added `DockerClientAuthenticationConfigurationTest`.
- Please keep the project related code styles in the future.

The next steps will be:

- [x] Add `DOCKER_TLS_VERIFY` and `DOCKER_CERT_PATH` from `DockerClientAuthConfig`. Apply them as default configuration. See `DockerClientAuthenticationConfiguration`. ~~Dunno the exact implementation yet~~.
- [x] Add interface to `ITestcontainersBuilder`. `WithDockerEndpoint(string clientEndpoint, bool tlsEnabled)` is just a placeholder. ~~Which primitive properties do we need~~?
- [x] We will extend `IDockerClientAuthenticationConfiguration` and add the necessary properties here.
- [ ] Add the client certificate credentials back again to `DockerApiClient`.
    - I think we can extend `CertificateCredentials` and add our implementation to load the certificates.
- [ ] Implement proper unit tests (`DockerClientAuthenticationConfigurationTest`).

Shouldn't be that much anymore.

### CODE REVIEW

File: src/DotNet.Testcontainers/Containers/Configurations/DockerClientEnvironmentAuthenticationConfiguration.cs [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+      catch (UriFormatException)
+      {
+        return null;
+      }
+    }
+
+    private static string GetDockerCertPathEnvVariable()
+    {
+      var dockerCertPath = Environment.GetEnvironmentVariable("DOCKER_CERT_PATH");
+      return string.IsNullOrEmpty(dockerCertPath) ? Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".docker") : dockerCertPath;
```
Comment (baez90) [inline src/DotNet.Testcontainers/Containers/Configurations/DockerClientEnvironmentAuthenticationConfiguration.cs:77]: I tried to find some docs about the certs directory on Windows and if I remember correctly it should be under `%HOMEDRIVE%%HOMEPATH%\.docker` but `Environment.GetFolderPath(Environment.SpecialFolder.UserProfile)` resolves to `%HOMEDRIVE%%HOMEPATH%\My Documents` - I'm almost 100% using Linux therefore I am not able to validate this and just wanted to ask if you're aware of this?

**Comment by baez90** [comment]
I'm really sorry that you had to work over this that intense. Of course it was not intended that you have to extra effort with this task!

Also sorry about the code styles, I tried to stick to your conventions where I noticed them, do you mind if I add some entries to the .editorconfig to enforce your style of brackets?

Probably I failed in expressing myself when I last commented on the initial PR but I really appreciate the way you updated the client configuration part!

Just a few questions:

> * Added `IDockerClientAuthenticationConfiguration`.

isn't this more a  `IDockerClientConfiguration`? Besides, the way you defined it only allows optional certificate authentication but does not cover e.g. HTTP basic auth, just to make sure this is by purpose?

**Comment by HofmeisterAn** [comment]
> Of course it was not intended that you have to extra effort with this task!

Oh no worries. It wasn't that much. You prepared a lot already; I just applied the pattern of how data gets passed through.

> isn't this more a `IDockerClientConfiguration`?

Yea, maybe a different name is better. I'm not sure. I was just looking for something that's not too similar to `IAuthenticationConfiguration` → `IDockerRegistryAuthenticationConfiguration`. I suggest we add the implementation first and discuss proper names later?

> Besides, the way you defined it only allows optional certificate authentication but does not cover e.g. HTTP basic auth, just to make sure this is by purpose?

Is there anything like HTTP basic authentication (against the Docker API)?

File: src/DotNet.Testcontainers/Containers/Configurations/DockerClientEnvironmentAuthenticationConfiguration.cs (hunk 1 above)
Comment (HofmeisterAn) [reply src/DotNet.Testcontainers/Containers/Configurations/DockerClientEnvironmentAuthenticationConfiguration.cs:77]: `Environment.GetFolderPath(Environment.SpecialFolder.UserProfile)` resolves to:

- macOS: `/Users/$USERNAME`.
- Windows: `C:/Users/$USERNAME`.

Comment (baez90) [reply src/DotNet.Testcontainers/Containers/Configurations/DockerClientEnvironmentAuthenticationConfiguration.cs:77]: Another round 'don't trust what you read on StackOverflow' - thanks for clarification!

Comment (HofmeisterAn) [reply src/DotNet.Testcontainers/Containers/Configurations/DockerClientEnvironmentAuthenticationConfiguration.cs:77]: Maybe it has changed in .NET Core. I have just tested both.

**Comment by baez90** [comment]
> Yea, maybe a different name is better. I'm not sure. I was just looking for something that's not too similar to IAuthenticationConfiguration → IDockerRegistryAuthenticationConfiguration. I suggest we add the implementation first and discuss proper names later?

totally agree on that - just wanted to know if I should stick to the names :blush: 

> Is there anything like HTTP basic authentication (against the Docker API)?

At least there's a prepared [NuGet package](https://www.nuget.org/packages/Docker.DotNet.BasicAuth/) for it therefore I thought it might be good to keep at least the option even if I don't know if there are a lot of users out there that add basic auth instead of cert auth but I found a few tutorials when I stumbled upon it and it's actually fairly easy to setup - hence I somehow expect someone to give it a try :sweat_smile:

**Comment by HofmeisterAn** [comment]
> At least there's a prepared [NuGet package](https://www.nuget.org/packages/Docker.DotNet.BasicAuth/) for it

Ok. I didn't know about that. Do you know if they use `Credentials` too (I guess they do)? Then it makes much more sense to replace `IsTlsVerificationEnabled`, `CertificatesDirectory` with `Credentials` and add dedicated implementations. We can keep `IsApplicable` to find the default implementation.

**Comment by baez90** [comment]
Yes they [do](https://github.com/dotnet/Docker.DotNet/blob/master/src/Docker.DotNet.BasicAuth/BasicAuthCredentials.cs) that's why I originally came up with the `Credentials` property even though I understand why you wanted to keep the API independent from the Docker client API.
Btw. that was also the original idea of the different static methods to allow different mechanisms by adding more 'factory' methods that allow to create different `Credentials` implementations.

I'll have a look in the upcoming days how to achieve all this with your approach of having the `IDockerClientAuthenticationConfiguration`.

**Comment by HofmeisterAn** [comment]
> Yes they do that's why I originally came up with the Credentials property

Ah ok, got it. We can simply add `Credentials` to the interface definition and add different implementations for client certificate, basic authentication, etc. Then `DockerClientEnvironmentAuthenticationConfiguration` would be split into separate implementations.

**Pull request merged**